### PR TITLE
Use Alarms API instead of intervals in automation

### PR DESCRIPTION
**Pull request by bershanskiy**
Previously, extension "automation" feature was checking every 10 seconds whether it was "night" or "day", even though typically night-day transition happens only 2 times in a 24 hour period. This commit enables extension to determine when exactly this transition is expected to happen next time and schedule an alarm for exact time.

This commit changes return value of `isInTimeInterval` and `isNightAtLocation` from simple boolen to object which contains the old boolean and a timestamp of when the answer will change. ~~To limit LoC and simplify code review, these functions schedule the new check for a dummy time, which replicates current extension behavior of running the check every 10 seconds.~~

### CODE REVIEW

File: src/utils/time.ts [hunk 1]
```diff
@@ -37,14 +39,24 @@ function compareTime(a: number[], b: number[]) {
     return 1;
 }
 
-export function isInTimeInterval(date: Date, time0: string, time1: string) {
+function dummyTime() {
```
Comment (bershanskiy) [inline src/utils/time.ts:42]: I will remove this `dummyTime` function in future commits once `isInTimeInterval` and `isNightAtLocation` can calculate exact time.

Comment (Gusted) [reply src/utils/time.ts:42]: For the first portion of converting: 
[patches.tar.gz](https://github.com/darkreader/darkreader/files/6874024/patches.tar.gz)

The always night/day time requires some more specialized Math. Otherwise we could just fallback to check it every day.

Comment (Gusted) [reply src/utils/time.ts:42]: The calculation for it would be more or less brute-forcing which is something I don't like so you could just fallback to check it every day, even though if someone was in this time-range they happen for around ~180 days before going back to the correct cycle. Which is another way to calculate the approximate.

Comment (bershanskiy) [reply src/utils/time.ts:42]: I excluded the actual calculation changes from this PR to simplify reviewing this PR, however I can add them in if you prefer.

> The calculation for it would be more or less brute-forcing which is something I don't like

As far as I see, calculating sunset/sunrise would be a brute-force for each _day_ of seasonal day/night, which is still better than brute-forcing every 10 seconds. So falling back to checking every day might not be that bad. However, I'd prefer to just fall back to scheduling the next alarm in 24 hours and defering this problem until later.

Comment (Gusted) [reply src/utils/time.ts:42]: > I excluded the actual calculation changes from this PR to simplify reviewing this PR, however I can add them in if you prefer.

Just add them now. As in both cases I will just approve them and merge them.

>  I'd prefer to just fall back to scheduling the next alarm in 24 hours and defering this problem until later.

👍🏽

Comment (bershanskiy) [reply src/utils/time.ts:42]: I just added time calculation for isInTimeInterval (with tests). Solar dawn/sunset is next.

File: src/utils/time.ts [hunk 2]
```diff
@@ -196,7 +196,7 @@ function getSunsetSunriseUTCTime(
         return {
             alwaysDay: false,
             alwaysNight: false,
-            time: UT * getDuration({hours: 1}),
+            time: Math.round(UT * getDuration({hours: 1})),
```
Comment (bershanskiy) [inline src/utils/time.ts:199]: This rounds number of milliseconds to the nearest integer, which is convenient for debugging and writing tests. This rounding does not actually lead to any perceived precision loss since JS event loop can introduce delays over a millisecond anyway.

File: src/utils/time.ts (hunk 1 above)
Comment (bershanskiy) [reply src/utils/time.ts]: I added support for calculating solar dawn/sunset and tests. The results make sense (they fall within reasonable ballpark), but do not match exactly any calculators I could find on the web. I suspect that these slight variations (within an hour) can be caused by different definitions of "dawn" and "sunset" in different models. E.g., someone might say dawn is when you see first sunrays creating ambient light, someone might say it is when sun appears on the horizon and someone might say it is when sun is above horizon and provides reasonable illumination.

**Comment by bershanskiy** [comment]
Apparently, tests are failing because two millisecond timing values differ between my PC and GitHub Actions runs. Specifically, 1568093140142 and 1568093042501. That is ~97 seconds of difference, which is a bit much.

**Comment by Gusted** [comment]
Even weirder: I'm getting a totally complete other error on another line...

image

**Comment by Gusted** [comment]
I have a good feeling it might be due to `UTC` being used everywhere which is something that doesn't go well, given that everyone lives in different time-zones and doesn't follow UTC time-zone and will actually receive incorrect timings due to it as they can live +8 hours away from UTC or -5 hours.

**Comment by bershanskiy** [comment]
> I have a good feeling it might be due to UTC being used everywhere

Time zones would have a much larger impact than 97 seconds. I suspect it is due to floating point precission loss somewhere in `getTime`, but I don't know how to find an exact line. As a work around, it's possible to round off a few significant digits in tests.

**Comment by Gusted** [comment]
> I suspect it is due to floating point precission loss somewhere in `getTime`

In that case we actually should return a precise Date object. As we don't want to round off some tests in order to look good.

**Comment by bershanskiy** [comment]
Note that your difference is `97642` and mine is `97641`.

**Comment by bershanskiy** [comment]
> > I suspect it is due to floating point precission loss somewhere in getTime
>
> In that case we actually should return a precise Date object. As we don't want to round off some tests in order to look good.

Could you elaborate? If problem is caused by numeric digit loss, it wouldn't matter whether return value is a plain number or the same number converted to a `Date`.

**Comment by Gusted** [comment]
> Could you elaborate? If problem is caused by numeric digit loss, it wouldn't matter whether return value is a plain number or the same number converted to a `Date`.

Well, I still find it hard to imagine that it's numeric digit loss as the float precision should only happen at around ~53 bits and should be able to handle the precise number up until 9007199254740991 (2^53 -1). Which means that the epoch times `1627419407399` should be perfectly able to be stored without loss of precision, also if such things would happen, it will probably be a known thing and some workaround would be known as it isn't uncommon to store and handle epoch times. 


Date is an internal javascript object, which could work around such issue if that's the case. But like I said, I still think it's due to the usage of UTC, which isn't a good representation of the current date of said user or machine.

**Comment by Gusted** [comment]
Anyway - What I'm trying to say, the current logic resolves around UTC time. However we shouldn't return UTC time back to the user as the user isn't in always in UTC - but if we do we should take into account the timezone difference with UTC to reflect that back into the alarm api. The amount of conversion between numbers could cause for imprecision or whatever the cause is.

**Comment by bershanskiy** [comment]
@Gusted I found the bug and fixed it, now CI passes.

File: src/utils/time.ts [hunk 3]
```diff
@@ -116,9 +116,9 @@ function getSunsetSunriseUTCTime(
     latitude: number,
     longitude: number,
 ) {
-    const dec31 = new Date(date.getUTCFullYear(), 0, 0);
```
Comment (bershanskiy) [inline src/utils/time.ts:83]: This is the buggy line. The problem is `new Date(...)` returns an object which internally keeps track of user's time zone. Instead, we can use `Date.UTC(...)` which just returns a timestamp in UTC GMT which does not depend on user's time zone.

**Comment by Gusted** [comment]
> @Gusted I found the bug and fixed it, now CI passes.

That's great, let's forget our past discussion about the theory's 😄.

File: tests/utils/time.tests.ts [hunk 4]
```diff
@@ -36,27 +36,27 @@ test('Duration', () => {
...
-    expect(isNightAtLocation(utcDate(2019, 8, 9, 18, 0), 52, 0)).toBe(false);
-    expect(isNightAtLocation(utcDate(2019, 8, 9, 20, 0), 52, 0)).toBe(true);
-    expect(isNightAtLocation(utcDate(2019, 8, 9, 23, 59), 52, 0)).toBe(true);
+    expect(isNightAtLocation(52, 0, utcDate(2019, 8, 9, 0, 0))).toEqual({rightNow: true, nextCheck: Date.UTC(2019, 8, 9, 5, 24, 2, 501)});
+    expect(isNightAtLocation(52, 0, utcDate(2019, 8, 9, 5, 0))).toEqual({rightNow: true, nextCheck: Date.UTC(2019, 8, 9, 5, 24, 2, 501)});
+    expect(isNightAtLocation(52, 0, utcDate(2019, 8, 9, 7, 0))).toEqual({rightNow: false, nextCheck: 1568053786448});
+    expect(isNightAtLocation(52, 0, utcDate(2019, 8, 9, 12, 0))).toEqual({rightNow: false, nextCheck: 1568053786448});
+    expect(isNightAtLocation(52, 0, utcDate(2019, 8, 9, 18, 0))).toEqual({rightNow: false, nextCheck: 1568053786448});
+    expect(isNightAtLocation(52, 0, utcDate(2019, 8, 9, 20, 0))).toEqual({rightNow: true, nextCheck: 1568093042501});
+    expect(isNightAtLocation(52, 0, utcDate(2019, 8, 9, 23, 59))).toEqual({rightNow: true, nextCheck: 1568093042501});
```
Comment (bershanskiy) [inline tests/utils/time.tests.ts:45]: I believe that, e.g., `Date.UTC(2019, 8, 9, 5, 24, 2, 501)` is easier to read than timestamp `1568006642501`. Should I convert all other timestamps to this format or should I just use simple numeric timestamps?

Comment (Gusted) [reply tests/utils/time.tests.ts:45]: Yeah go for the read-ability

File: src/background/extension.ts [hunk 5]
```diff
@@ -54,30 +55,49 @@ export class Extension {
...
-            return isSystemDarkModeEnabled();
-        } else if (automation === 'location') {
-            const latitude = this.user.settings.location.latitude;
-            const longitude = this.user.settings.location.longitude;
-
-            if (latitude != null && longitude != null) {
-                const now = new Date();
-                return isNightAtLocation(now, latitude, longitude);
+
+        let goodTime: TimeCheck = null;
```
Comment (Gusted) [inline src/background/extension.ts:71]:
```suggestion
        let timingInformation: TimeCheck = null;
```

Or any other variable 😄

File: src/utils/time.ts [hunk 6]
```diff
@@ -27,24 +29,58 @@ function parse24HTime(time: string) {
...
-    if (a[0] < b[0] || (a[0] === b[0] && a[1] < b[1])) {
+    if (time1[0] < time2[0] || (time1[0] === time2[0] && time1[1] < time2[1])) {
         return -1;
     }
     return 1;
 }
 
-export function isInTimeInterval(date: Date, time0: string, time1: string) {
+// a <= b
+function nextIntervalTime(a, b, t, date: Date): number {
```
Comment (Gusted) [inline src/utils/time.ts:43]:
```suggestion
function nextIntervalTime(time1, time2, currentTime, date: Date): number {
```

File: tests/utils/time.tests.ts (hunk 4 above)
Comment (bershanskiy) [reply tests/utils/time.tests.ts]: Done!

**Review by Gusted — CHANGES_REQUESTED** [review]
Just some variable naming. The code on it's own and the logic is approved for me.

@alexanderby If you'd like to have a look soon 🤷🏽‍♀️, as this one will be merged soon.

File: src/background/extension.ts (hunk 5 above)
Comment (bershanskiy) [reply src/background/extension.ts]: Done!

File: src/utils/time.ts (hunk 6 above)
Comment (bershanskiy) [reply src/utils/time.ts]: Done!

File: src/utils/time.ts [hunk 7]
```diff
@@ -204,8 +247,70 @@ export function isNightAtLocation(
...
     );
 
-    if (sunsetTime > sunriseTime) {
-        return (currentTime > sunsetTime) || (currentTime < sunriseTime);
+    if (sunriseTime < sunsetTime) {
+        // Timeline:
+        // --- sunrise <----> sunset ---
+        let rightNow: boolean;
+        let nextCheck: number;
+        if (currentTime < sunriseTime) {
```
Comment (Gusted) [inline src/utils/time.ts:255]: What about moving the if statements a bit around and simplify it a bit for us.

```ts
if ((sunriseTime < currentTime) && (currentTime < sunsetTime)) {
            // Timeline:
            // --- sunrise <----> sunset ---
            //               ^
            //          Current time
            rightNow = false;
            nextCheck = Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate(), 0, 0, 0, sunsetTime);
} else {
            // Timeline:
            // --- sunrise <----> sunset ---
            //   ^                       ^
            //           Current time
            rightNow = true;
            nextCheck = Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate() + (sunsetTime < currentTime ? 1 : 0), 0, 0, 0, sunriseTime);
}
```

File: src/utils/time.ts [hunk 8]
```diff
@@ -204,8 +247,70 @@ export function isNightAtLocation(
...
+        return {
+            rightNow,
+            nextCheck,
+        };
+    }
+    // Timeline:
+    // --- sunset <----> sunrise ---
+    let rightNow: boolean;
+    let nextCheck: number;
+    if (currentTime < sunsetTime) {
```
Comment (Gusted) [inline src/utils/time.ts:288]: Same story for these functions.

File: src/utils/time.ts [hunk 9]
```diff
@@ -204,8 +247,70 @@ export function isNightAtLocation(
...
+    if ((sunsetTime < currentTime) && (currentTime < sunriseTime)) {
+        // Timeline:
+        // --- sunset <----> sunrise ---
+        //               ^
+        //          Current time
+        rightNow = true;
+        nextCheck = Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate(), 0, 0, 0, sunriseTime);
     }
-    return (currentTime > sunsetTime) && (currentTime < sunriseTime);
+    if ((sunsetTime < currentTime) && (currentTime < sunriseTime)) {
```
Comment (Gusted) [inline src/utils/time.ts:304]: Wrong if condition btw. But should be solved if we go `if () {} else {]`

**Review by Gusted — CHANGES_REQUESTED** [review]
Woops just a couple more 😃

File: src/background/extension.ts [hunk 10]
```diff
@@ -54,30 +55,49 @@ export class Extension {
...
+                }
+                break;
             }
+            default:
+                return this.user.settings.enabled;
         }
-
-        return this.user.settings.enabled;
+        this.isEnabledNow = timingInformation.rightNow;
+        if (timingInformation.nextCheck) {
```
[the remark this answers is not in the document]
Comment (bershanskiy) [reply src/background/extension.ts:97]: This is an unusual/edge case when time automation is used and start time is equal to end time. As of now, UI allows this to happen, but I feel like it shouldn't.

This behavior is described by the following `isInTimeInterval` tests:
```
expect(isInTimeInterval('10:00', '10:00', new Date(2018, 11, 4, 10))).toEqual({rightNow: false, nextCheck: null});
expect(isInTimeInterval('10:00', '10:00', new Date(2018, 11, 4, 12))).toEqual({rightNow: false, nextCheck: null});
expect(isInTimeInterval('10:00', '10:00', new Date(2018, 11, 4, 8))).toEqual({rightNow: false, nextCheck: null});
```

File: src/background/extension.ts [hunk 11]
```diff
@@ -54,30 +55,49 @@ export class Extension {
...
 
-    isEnabled() {
+    private alarmListener = (alarm): void => {
+        if (alarm.name === Extension.ALARM_NAME) {
+            this.handleAutoCheck();
+        }
+    };
+
+    isEnabled(): boolean {
+        if (this.isEnabledNow !== null) {
```
[the remark this answers is not in the document]
Comment (bershanskiy) [reply src/background/extension.ts:65]: I'm using `null` on purpose here. Removing it would modify existing logic for case when `isEnabledNow` is `false`.

**Review by Gusted — APPROVED** [review]
All seems good for me. If you think it's ready let me know so I can merge it.

**Comment by bershanskiy** [comment]
Yes, I think this is ready now.

**Comment by Gusted** [comment]
Thank you 😃

File: src/background/extension.ts [hunk 12]
```diff
@@ -33,6 +31,9 @@ export class Extension {
     tabs: TabManager;
     user: UserStorage;
 
+    private isEnabledNow: boolean = null;
```
Comment (alexanderby) [inline src/background/extension.ts:34]: Maybe you could remove `isEnabledNow`? Is it an optimisation? I think it makes code harder to understand and to keep track of its state. Or at least remove to something like `isEnabledByAutomation`.

File: src/background/extension.ts [hunk 13]
```diff
@@ -33,6 +31,9 @@ export class Extension {
...
+
+    static ALARM_NAME = 'auto-time-alarm';
```
Comment (alexanderby) [inline src/background/extension.ts:36]: `ALARM_NAME` seems a generic name.

As an idea, maybe try using 2 alarms? One for the beginning of time interval and another one for the ending?

File: src/utils/time.ts [hunk 14]
```diff
@@ -27,24 +29,57 @@ function parse24HTime(time: string) {
...
+        date.setMinutes(time2[1]);
+        return date.getTime();
+    }
+
+    // a <= b <= t
+    // Schedule for tomorrow at time a
+    return (new Date(date.getFullYear(), date.getMonth(), date.getDate() + 1, time1[0], time1[1])).getTime();
+}
+
+export function isInTimeInterval(time0: string, time1: string, date: Date = new Date()): TimeCheck {
```
Comment (alexanderby) [inline src/utils/time.ts:69]: `isInTimeInterval` should only return a boolean value, it should not know anything about next intervals etc.

Try exporting `nextIntervalTime` function, but make its name understandable out of context.

**Review by alexanderby — COMMENTED** [review]
Sorry for the delay, I was feeling bad these days.

**Comment by alexanderby** [comment]
@bershanskiy By the way, could you please check your email?

File: src/background/extension.ts (hunk 12 above)
Comment (bershanskiy) [reply src/background/extension.ts:34]: Yes, `isEnabledNow` is just an optimisation for avoiding to evaluate these interval checks repeatedly. This is still useful for `getTabMessage()`:

https://github.com/darkreader/darkreader/blob/0301db28e7e6c461033d56dd18c8b47d609b8704/src/background/extension.ts#L414-L417

As far as I know, when user toggles extension ON, `getTabMessage` calls `isEnabled()` for every tab, which might be a lot.

File: src/background/extension.ts (hunk 13 above)
Comment (bershanskiy) [reply src/background/extension.ts:36]: > `ALARM_NAME` seems a generic name.

I had to create internal alarms for Newsmaker, Network, and Extension classes, so I called them `Newsmaker.ALARM_NAME`, `LimitedCacheStorage.ALARM_NAME`, and `Extension.ALARM_NAME` as a naming convention. If you have a better suggestion, I can rename them and one in #6327. (I just need static string variables with unique names, but their names and even values can be entirely arbitrary.)

> As an idea, maybe try using 2 alarms? One for the beginning of time interval and another one for the ending?

I'm not sure how that would be an improvement, could you elaborate? These alarms can't be recurring alarms anyway because the exact sunset/dawn times vary a bit.

File: src/utils/time.ts (hunk 14 above)
Comment (bershanskiy) [reply src/utils/time.ts:69]: Makes sense, will prepare a PR today or tomorrow.

File: src/background/extension.ts (hunk 12 above)
Comment (bershanskiy) [reply src/background/extension.ts:34]: I'll bundle this rename together with `isInTimeInterval` PR.

File: src/utils/time.ts (hunk 14 above)
Comment (bershanskiy) [reply src/utils/time.ts:69]: PR is at #6405.
